addons/account/models/account_move.py

```diff
@@ -3197,8 +3218,40 @@ def field_will_change(line, field_name):
                             or (account_type != 'payable' and account_to_write.user_type_id.type == 'payable'):
                         raise UserError(_("You can only set an account having the payable type on payment terms lines for vendor bill."))
 
+        # Tracking enable only for posted moves even there are reset to draft.
+        for line in self.filtered(lambda l: l.move_id.name != '/'):
+            # Get tracking value and set a dict with some information for the message post.
+            for value in vals:
+                field = line._fields[value]
+                # If the field is related, we don't want to track the field in this model.
+                if hasattr(field, 'related') and field.related:
+                    continue
+                # The field must be tracked and changed !
+                if field_will_change(line, value) and hasattr(field, 'tracking') and field.tracking:
+                    modified_field = {
+                        'field': value,
+                        'field_name': field.string,
+                        'move_id': line.move_id.id,
+                        'old_value': get_tracking_field_values(field, line[value]),
+                        'new_value': None, # This information is calculated after the account move line is written.
+                    }
+                    tracked_modified_fields.append(modified_field)
+
         result = super(AccountMoveLine, self).write(vals)
 
+        # Tracking enable only for posted moves even there are reset to draft.
+        for line in self.filtered(lambda l: l.move_id.name != '/'):
+            for modified_field in tracked_modified_fields:
+                field = line._fields[modified_field.get('field')]
+                value = line[modified_field.get('field')] # Get the new value
+                modified_field.update({'new_value': get_tracking_field_values(field, value)}) # Get the well formated new value
```

Comment:
using the proposed structure will allow an easier loop on objects.
Also try to reuse some of the functions of mail_thread. I think that env[account.move.line]._message_track() should be reusable.

``` python
for move, initial_values in moves_initial_values.items():
    for line, line_initial_value in intial_values.items():
        changes, tracking_value_ids =  line._message_track(tracked_fields, line_initial_value)
        ...
    if changes:
        move.message_post()
```